runtime-light/stdlib/rpc/rpc-api.cpp

```diff
@@ -0,0 +1,522 @@
+  }
+  const auto query_id{rpc_ctx.current_query_id++};
+  rpc_ctx.pending_component_queries.emplace(query_id, std::move(comp_query));
+  co_return RpcQueryInfo{.id = query_id, .request_size = request_size, .timestamp = timestamp};
+}
+
+task_t<RpcQueryInfo> rpc_tl_query_one_impl(const string &actor, mixed tl_object, double timeout, bool collect_resp_extra_info, bool ignore_answer) noexcept {
+  auto &rpc_ctx{RpcComponentContext::get()};
+
+  if (!tl_object.is_array()) {
+    rpc_ctx.current_query.raise_storing_error("not an array passed to function rpc_tl_query");
+    co_return RpcQueryInfo{};
+  }
+
+  rpc_ctx.buffer.clean();
+  auto rpc_tl_query{store_function(tl_object)}; // TODO: exception handling
+  if (rpc_tl_query.is_null()) {
+    co_return RpcQueryInfo{};
+  }
+
+  const auto query_info{co_await rpc_send_impl(actor, timeout, ignore_answer)};
+  if (!ignore_answer) {
+    rpc_ctx.pending_rpc_queries.emplace(query_info.id, std::move(rpc_tl_query));
+  }
+  if (collect_resp_extra_info) {
+    rpc_ctx.rpc_responses_extra_info.emplace(query_info.id,
+                                             std::make_pair(rpc_response_extra_info_status_t::NOT_READY, rpc_response_extra_info_t{0, query_info.timestamp}));
+  }
+
+  co_return query_info;
+}
+
+task_t<RpcQueryInfo> typed_rpc_tl_query_one_impl(const string &actor, const RpcRequest &rpc_request, double timeout, bool collect_responses_extra_info,
+                                                 bool ignore_answer) noexcept {
+  auto &rpc_ctx{RpcComponentContext::get()};
+
+  if (rpc_request.empty()) {
+    rpc_ctx.current_query.raise_storing_error("query function is null");
+    co_return RpcQueryInfo{};
+  }
+
+  rpc_ctx.buffer.clean();
+  auto fetcher{rpc_request.store_request()};
+  if (!static_cast<bool>(fetcher)) {
+    rpc_ctx.current_query.raise_storing_error("could not store rpc request");
+    co_return RpcQueryInfo{};
+  }
+
+  const auto query_info{co_await rpc_send_impl(actor, timeout, ignore_answer)};
+  if (!ignore_answer) {
+    auto rpc_tl_query{make_instance<RpcTlQuery>()};
+    rpc_tl_query.get()->result_fetcher = std::move(fetcher);
+    rpc_tl_query.get()->tl_function_name = rpc_request.tl_function_name();
+
+    rpc_ctx.pending_rpc_queries.emplace(query_info.id, std::move(rpc_tl_query));
+  }
+  if (collect_responses_extra_info) {
+    rpc_ctx.rpc_responses_extra_info.emplace(query_info.id,
+                                             std::make_pair(rpc_response_extra_info_status_t::NOT_READY, rpc_response_extra_info_t{0, query_info.timestamp}));
+  }
+
+  co_return query_info;
+}
+
+task_t<array<mixed>> rpc_tl_query_result_one_impl(int64_t query_id) noexcept {
+  if (query_id < RPC_VALID_QUERY_ID_RANGE_START) {
+    co_return make_fetch_error(string{"wrong query_id"}, TL_ERROR_WRONG_QUERY_ID);
+  }
+
+  auto &rpc_ctx{RpcComponentContext::get()};
+  class_instance<RpcTlQuery> rpc_query{};
+  class_instance<C$ComponentQuery> component_query{};
+
+  {
```

Comment:
Is nested scope here necessary?

Reply:
It's as necessary as `RAII` is necessary :) Sure, we can go without that nested scope. But I use it to make sure that there will be no error when someone tries to use iterators to internal maps, e.g. `rpc_ctx.pending_rpc_queries`